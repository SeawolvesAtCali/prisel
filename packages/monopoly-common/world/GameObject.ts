import { deserialize, serializable, serialize } from 'serializr';
import { exist } from '../exist';
import { StaticThis } from './staticThis';
import { World } from './World';

export abstract class GameObject {
    /**
     * type is classname. This is used during (de)serialization to identify the
     * class of game object.
     */
    public abstract get type(): string;
    /**
     * Id uniquely identify a game object
     */
    @serializable
    public id: string = '';

    private _world?: World;
    public get world() {
        if (exist(this._world)) {
            return this._world;
        }
        throw new Error(`GameObject with id ${this.id} doesn't have a world assigned`);
    }
    public set world(world: World) {
        this._world = world;
    }

    // Child type of GameObject should all provide a no argument constructor, or
    // just don't add constructor. This allows world to create any GameObject

    public serialize() {
        return serialize(this);
    }

    public static deserialize<T extends GameObject>(
        this: StaticThis<T>,
        serialized: any,
        world: World,
    ) {
        const deserialized = deserialize(this, serialized, undefined, { world });
        world.add(deserialized);
        return deserialized;
    }
    /**
     * Return true if all required mixins are initialized
     */
    public check(): boolean {
        return true;
    }
}

export type GameObjectClass<T> = T extends GameObject
    ? {
          new (): T;
          TYPE: string;
          deserialize(serialized: any, world: World): T;
      }
    : never;
