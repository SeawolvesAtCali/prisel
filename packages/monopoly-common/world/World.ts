import { serialize, update } from 'serializr';
import { genId } from '../genId';
import { GameObject, GameObjectClass } from './GameObject';
import { Id } from './Id';
import { Ref as Ref2 } from './ref2';

export interface SerializedWorld {
    [key: string]: Array<any>;
}

/**
 * World holds all game objects. Provide easy access to game object through ids
 */
export class World {
    private objectMap: Map<Id, GameObject> = new Map();
    private idMap: Map<string, Set<Id>> = new Map();

    constructor() {
        this.makeRef = this.makeRef.bind(this);
    }

    private createId<T extends GameObject>(specifiedId?: Id<T>): Id<T> {
        if (specifiedId && this.objectMap.has(specifiedId)) {
            throw new Error(
                `Cannot create game object of id ${specifiedId} because there is already one existing`,
            );
        }
        if (specifiedId) {
            return specifiedId;
        }
        return genId();
    }

    public get<T extends GameObject>(id: Id<T>): T | null {
        const object = this.objectMap.get(id);
        return (object as T) || null;
    }

    private getIdSetOfSameType(object: GameObject) {
        const type = object.type;
        // const gameObjectClass = this.gameObjectTypeRegistry.get(type);
        // if (gameObjectClass && object instanceof gameObjectClass) {
        const set = this.idMap.get(type);
        if (!set) {
            const newSet = new Set<string>();
            this.idMap.set(type, newSet);
            return newSet;
        }
        return set;
    }

    private getAllFromSet<T extends GameObject>(idSet: Set<Id>): T[] {
        return Array.from(idSet)
            .map((id) => this.objectMap.get(id) as T)
            .filter(Boolean);
    }

    public getAll<T extends GameObjectClass<any>>(typeOrClazz: string | T): InstanceType<T>[] {
        const typeString = typeof typeOrClazz === 'string' ? typeOrClazz : typeOrClazz.TYPE;
        const set = this.idMap.get(typeString);
        if (set) {
            return this.getAllFromSet<InstanceType<T>>(set);
        }
        return [];
    }

    /** Create a GameObject of given type, and add to world */
    public create<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): InstanceType<T> {
        const object = new clazz();
        object.world = this;
        const objectId = this.createId<InstanceType<T>>(id);
        object.id = objectId;
        this.objectMap.set(objectId, object);
        this.getIdSetOfSameType(object).add(objectId);

        return object;
    }

    public createRef<T extends GameObjectClass<any>>(
        clazz: T,
        id?: Id<InstanceType<T>>,
    ): Ref2<InstanceType<T>> {
        return this.makeRef(this.create(clazz, id));
    }

    public add<T extends GameObject>(object: T) {
        if (this.idMap.get(object.type)?.has(object.id)) {
            throw new Error(`object ${object.type} already has an id ${object.id}`);
        }
        this.objectMap.set(object.id, object);
        this.getIdSetOfSameType(object).add(object.id);
        object.world = this;
    }

    public remove<T extends GameObject>(idOrGameObject: Id | T) {
        let deletedObject;
        if (typeof idOrGameObject === 'string') {
            deletedObject = this.objectMap.get(idOrGameObject);
            this.objectMap.delete(idOrGameObject);
        } else {
            deletedObject = idOrGameObject;
            this.objectMap.delete(idOrGameObject.id);
        }
        if (deletedObject) {
            this.getIdSetOfSameType(deletedObject).delete(deletedObject.id);
        }
    }

    public makeRef<T extends GameObject>(keyOrObject: Id<T> | T): Ref2<T> {
        return Ref2.of<T>(keyOrObject, this);
    }

    /**
     * Called before serializing. This method should be used to populate the
     * serializable fields in sub class.
     */
    protected populateSerializedFields(): void {}
    public serialize(): any {
        this.populateSerializedFields();
        const serialzied = serialize(this);
        this.clearSerializedFields();
        return serialzied;
    }

    /**
     * Called after deserialized. This method should be used to clear out the
     * serialized fields to reduce duplicate reference. All GameObjects should
     * already be stored inside the internal map of World.
     */
    protected clearSerializedFields(): void {}
    /**
     * deserialize into current world make sure this is an empty world,
     * otherwise we might have id collision
     * @param serialized
     */
    public populate(serialized: any) {
        this.clearSerializedFields();
        update(
            this,
            serialized,
            (err) => {
                if (err) {
                    console.error(err);
                } else {
                    this.clearSerializedFields();
                }
            },
            { world: this },
        );
        return this;
    }
}
