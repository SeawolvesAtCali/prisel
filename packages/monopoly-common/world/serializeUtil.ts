import { custom, list, raw as _raw, serializable } from 'serializr';
import { exist } from '../exist';
import { GameObject, GameObjectClass } from './GameObject';
import { Ref } from './ref2';

const refPropSchema = custom(
    (ref: Ref<any>) => (exist(ref) ? ref.id : null),
    (jsonValue, context) =>
        jsonValue === null ? undefined : Ref.of(`${jsonValue}`, context.args.world),
);
export const refSerializable = serializable(refPropSchema);
export const listRefSerializable = serializable(list(refPropSchema));

// A simple serializer that serialize into json string
export const raw = serializable(_raw());

const getGameObjectPropSchema = (clazz: GameObjectClass<any>) =>
    custom(
        (gameObject: GameObject) => gameObject.serialize(),
        (jsonValue, context) =>
            jsonValue == null ? undefined : clazz.deserialize(jsonValue, context.args.world),
    );

export const gameObjectSerializable = (clazz: GameObjectClass<any>) =>
    serializable(getGameObjectPropSchema(clazz));

export const listGameObjectSerizable = (clazz: GameObjectClass<any>) =>
    serializable(list(getGameObjectPropSchema(clazz)));
