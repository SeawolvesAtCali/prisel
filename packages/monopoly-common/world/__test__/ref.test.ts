import { GameObject } from '../GameObject';
import { Ref } from '../ref2';
import { listGameObjectSerizable, refSerializable } from '../serializeUtil';
import { World } from '../World';

class TestObject extends GameObject {
    static TYPE = 'test';
    readonly type = 'test';
    @refSerializable
    ref?: Ref<TestObject>;
}

class TestWorld extends World {
    @listGameObjectSerizable(TestObject)
    testObjects: TestObject[] = [];

    // override
    populateSerializedFields() {
        this.testObjects = this.getAll(TestObject);
    }

    // override
    clearSerializedFields() {
        this.testObjects = [];
    }
}

describe('ref', () => {
    test('serialize', () => {
        const world = new TestWorld();
        const object1 = world.create(TestObject, 'object1');
        const object2Ref = world.createRef(TestObject, 'object2');
        object1.ref = object2Ref;
        expect(object1.serialize()).toMatchSnapshot();
    });
    test('deserialize restores ref', () => {
        const world = new TestWorld();
        const object1 = world.create(TestObject, 'object1');
        const object2Ref = world.createRef(TestObject, 'object2');
        object1.ref = object2Ref;

        const serializedWorld = world.serialize();
        const newWorld = new TestWorld().populate(serializedWorld);
        expect(newWorld.get<TestObject>('object1')?.ref?.get()).toBe(newWorld.get('object2'));
    });
});
