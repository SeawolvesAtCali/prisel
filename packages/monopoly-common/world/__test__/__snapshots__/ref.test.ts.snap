// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ref serialize 1`] = `
Object {
  "id": "object1",
  "ref": "object2",
}
`;
